BUG: Flexible array member warning in C++ using 2.0.0b1 headers

### Describe the issue:

Thanks for all the work on NumPy 2.0. I have just run a few tests using 2.0.0b1 on a C++ extension that uses the NumPy array API and everything seems to build quite smoothly. 

I do get a new warning from `PyArrayMultiIterObject` (which I actually don't use) relating to its new use of a flexible array member [at the end of the struct](https://github.com/numpy/numpy/blob/acef1f61428288f2bbb75dc9e97fc016762c0084/numpy/_core/include/numpy/ndarraytypes.h#L1304) reporting that these aren't supported in ISO C++.

Granted, this is a warning from GCC's `-Wpedantic` and so is particularly picky, but the public headers for the 1.x versions did build in C++ even with these turned on. I took a quick glance and it looks like this might be the only flexible array member in public headers. I haven't dug in to the context around the change but I wanted to report this since it is the only warning I see. Thanks again!

### Reproduce the code example:

```python
// Could compile an empty C++ extension (for example, the below just including headers)
// set -Wpedantic with GCC

#define NPY_NO_DEPRECATED_API NPY_2_0_API_VERSION
#define NPY_TARGET_VERSION NPY_2_0_API_VERSION

#include <Python.h>
#include <numpy/arrayobject.h>
```


### Error message:

```shell
venv/lib/python3.12/site-packages/numpy/_core/include/numpy/ndarraytypes.h:1304:31: warning: ISO C++ forbids flexible array member 'iters' [-Wpedantic]
```


### Python and NumPy Versions:

NumPy 2.0.0b1
Python 3.12.0

### Runtime Environment:

_No response_

### Context for the issue:

Leads to added warnings when building as C++ and relies on compiler extensions

--- a/numpy/_core/include/numpy/ndarraytypes.h
+++ b/numpy/_core/include/numpy/ndarraytypes.h
@@ -1298,9 +1298,11 @@ typedef struct {
          * growing structs (as of Cython 3.0.6).  It also allows NPY_MAXARGS
          * to be runtime dependent.
          */
-#if defined(NPY_INTERNAL_BUILD) && NPY_INTERNAL_BUILD
-        PyArrayIterObject    *iters[64];  /* 64 is NPY_MAXARGS */
-#else /* not internal build */
+#if (defined(NPY_INTERNAL_BUILD) && NPY_INTERNAL_BUILD) || defined(__cplusplus)
+        /* 64 is NPY_MAXARGS for numpy 2.0 or newer. We can't use a flexible
+           array member in C++ so use the internal size there. */
+        PyArrayIterObject    *iters[64];
+#else
         PyArrayIterObject    *iters[];
 #endif
 } PyArrayMultiIterObject;

